File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    let l = 0;
    let r = 0;
    let max = 0;
    const aCharCode = 'A'.charCodeAt(0);
    const zCharCode = 'Z'.charCodeAt(0);
    const chars = new Array(zCharCode - aCharCode + 1).fill(0);

    for(; r < s.length; r++) {
        let idx = s.charCodeAt(r) - aCharCode;
        chars[idx]++;
        max = Math.max(max, chars[idx]);

        if(r - l + 1 - max > k) {
            chars[s.charCodeAt(l) - aCharCode]--;
            l++;
        }
    }

    return r - l;
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int l = 0;
        int max = 0;
        int r = 0;
        
        final int aCode = Character.codePointAt("A", 0);
        final int zCode = Character.codePointAt("Z", 0);
        
        int arraySize = zCode - aCode + 1;
        int[] chars = new int[arraySize];
        java.util.Arrays.fill(chars, 0);
        
        for(; r < s.length(); r++){
            int idx = Character.codePointAt(s,r) - aCode;
            chars[idx]++;
            max = Math.max(max, chars[idx]);
            
            if(r-l+1-max > k){
                chars[Character.codePointAt(s,l) - aCode]--;
                l++;
            }
        }
        return r-l;
    }
}